controller/controller.go

```diff
@@ -0,0 +1,210 @@
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Reconciler is the interface that controller implementations are expected
+// to implement, so that the shared controller.Impl can drive work through it.
+type Reconciler interface {
+	Reconcile(ctx context.Context, key string) error
+}
+
+// PassNew makes it simple to create an UpdateFunc for use with
+// cache.ResourceEventHandlerFuncs that can delegate the same methods
+// as AddFunc/DeleteFunc but passing through only the second argument
+// (which is the "new" object).
+func PassNew(f func(interface{})) func(interface{}, interface{}) {
+	return func(first, second interface{}) {
+		f(second)
+	}
+}
+
+// Filter makes it simple to create FilterFunc's for use with
+// cache.FilteringResourceEventHandler that filter based on the
+// schema.GroupVersionKind of the controlling resources.
+func Filter(gvk schema.GroupVersionKind) func(obj interface{}) bool {
+	return func(obj interface{}) bool {
+		if object, ok := obj.(metav1.Object); ok {
+			owner := metav1.GetControllerOf(object)
+			return owner != nil &&
+				owner.APIVersion == gvk.GroupVersion().String() &&
+				owner.Kind == gvk.Kind
+		}
+		return false
+	}
+}
+
+// Impl is our core controller implementation.  It handles queuing and feeding work
+// from the queue to an implementation of Reconciler.
+type Impl struct {
+	// Reconciler is the workhorse of this controller, it is fed the keys
+	// from the workqueue to process.  Public for testing.
+	Reconciler Reconciler
+
+	// WorkQueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens. This
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	WorkQueue workqueue.RateLimitingInterface
+
+	// Sugared logger is easier to use but is not as performant as the
+	// raw logger. In performance critical paths, call logger.Desugar()
+	// and use the returned raw logger instead. In addition to the
+	// performance benefits, raw logger also preserves type-safety at
+	// the expense of slightly greater verbosity.
+	logger *zap.SugaredLogger
```

Comment:
should this be a logger or a context?

Reply:
I don't think so?  I am assuming this is in reference to my comment on the other PR?